// @ngstack/code-editor/esm2020/lib/code-editor.module.mjs
import { NgModule, APP_INITIALIZER } from '@angular/core';
import { CommonModule } from '@angular/common';
import { CodeEditorComponent } from './code-editor/code-editor.component';
import { CodeEditorService, EDITOR_SETTINGS, } from './services/code-editor.service';
import { TypescriptDefaultsService } from './services/typescript-defaults.service';
import { JavascriptDefaultsService } from './services/javascript-defaults.service';
import { JsonDefaultsService } from './services/json-defaults.service';
import * as i0 from "@angular/core";
export function setupEditorService(service) {
    const result = () => service.loadEditor();
    return result;
}
export class CodeEditorModule {
    static forRoot(settings) {
        return {
            ngModule: CodeEditorModule,
            providers: [
                { provide: EDITOR_SETTINGS, useValue: settings },
                CodeEditorService,
                TypescriptDefaultsService,
                JavascriptDefaultsService,
                JsonDefaultsService,
                {
                    provide: APP_INITIALIZER,
                    useFactory: setupEditorService,
                    deps: [CodeEditorService],
                    multi: true,
                },
            ],
        };
    }
    static forChild() {
        return {
            ngModule: CodeEditorModule,
        };
    }
}
CodeEditorModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.1", ngImport: i0, type: CodeEditorModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
CodeEditorModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.1", ngImport: i0, type: CodeEditorModule, declarations: [CodeEditorComponent], imports: [CommonModule], exports: [CodeEditorComponent] });
CodeEditorModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.1", ngImport: i0, type: CodeEditorModule, imports: [CommonModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.1", ngImport: i0, type: CodeEditorModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    declarations: [CodeEditorComponent],
                    exports: [CodeEditorComponent],
                }]
        }] });
